cmd/s3-peer-client.go

```diff
@@ -19,125 +19,74 @@ package cmd
-	globalS3Peers = s3Peers{
-		rpcClients: make(map[string]*AuthRPCClient),
-		mutex:      &sync.RWMutex{},
-	}
+	// Setup local server state change handler
+	bmsc[globalMinioAddr] = &localBMS{ObjectAPI: newObjectLayerFn}
 
-	// Initialize each peer connection.
-	for _, peer := range peers {
-		globalS3Peers.InitS3PeerClient(peer)
-	}
+	peers := []string{globalMinioAddr}
 
-	// Save new peers
-	globalS3Peers.peers = peers
-
-	// store if this is a distributed setup or not.
-	globalS3Peers.isDistXL = len(globalS3Peers.peers) > 1
-}
+	for _, ep := range eps {
+		if ep == nil || ep.Host == "" || ep.Host == globalMinioAddr {
+			continue
+		}
 
-func (s3p *s3Peers) GetPeers() []string {
-	return s3p.peers
-}
+		// Check if the remote host has been added already
+		if _, ok := bmsc[ep.Host]; !ok {
+			cfg := authConfig{
+				accessKey:   serverConfig.GetCredential().AccessKeyID,
+				secretKey:   serverConfig.GetCredential().SecretAccessKey,
+				address:     ep.Host,
+				secureConn:  isSSL(),
+				path:        path.Join(reservedBucket, s3Path),
+				loginMethod: "S3.LoginHandler",
+			}
 
-func (s3p *s3Peers) GetPeerClient(peer string) *AuthRPCClient {
-	// Take a read lock
-	s3p.mutex.RLock()
-	defer s3p.mutex.RUnlock()
-	return s3p.rpcClients[peer]
-}
+			bmsc[ep.Host] = &remoteBMS{newAuthClient(&cfg)}
 
-// Initializes a new RPC connection (or closes and re-opens if it
-// already exists) to a peer. Note that peer address is in `host:port`
-// format.
-func (s3p *s3Peers) InitS3PeerClient(peer string) {
-	// Take a write lock
-	s3p.mutex.Lock()
-	defer s3p.mutex.Unlock()
-
-	if s3p.rpcClients[peer] != nil {
-		_ = s3p.rpcClients[peer].Close()
-		delete(s3p.rpcClients, peer)
+			peers = append(peers, ep.Host)
+		}
 	}
-	authCfg := &authConfig{
-		accessKey:   serverConfig.GetCredential().AccessKeyID,
-		secretKey:   serverConfig.GetCredential().SecretAccessKey,
-		address:     peer,
-		secureConn:  isSSL(),
-		path:        path.Join(reservedBucket, s3Path),
-		loginMethod: "S3.LoginHandler",
+
+	return s3Peers{
+		bmsClients: bmsc,
+		IsDistXL:   len(peers) > 1,
+		Peers:      peers,
 	}
-	s3p.rpcClients[peer] = newAuthClient(authCfg)
 }
 
-func (s3p *s3Peers) Close() error {
-	// Take a write lock
-	s3p.mutex.Lock()
-	defer s3p.mutex.Unlock()
-
-	for _, v := range s3p.rpcClients {
-		if err := v.Close(); err != nil {
-			return err
-		}
-	}
-	s3p.rpcClients = nil
-	s3p.peers = nil
-	return nil
+func initGlobalS3Peers(eps []*url.URL) {
+	globalS3Peers = mkPeersFromEndpoints(eps)
 }
 
-// Returns the network addresses of all Minio servers in the cluster in `host:port` format.
-func getAllPeers(eps []*url.URL) (peers []string) {
-	if eps == nil {
-		return nil
-	}
-	peers = []string{globalMinioAddr} // Starts with a default peer.
-	for _, ep := range eps {
-		if ep == nil {
-			return nil
-		}
-		// Rest of the peers configured.
-		peers = append(peers, ep.Host)
-	}
-	return peers
+func (s3p *s3Peers) GetPeerClient(peer string) BucketMetaState {
```

Comment:
GetPeerClient() isn't adding value here, i.e if this was a library package and we did not want to expose s3p.bmsClients then GetPeerClient() could give us the control. Since `s3Peers` gets used by the same package, you can use s3p.bmsClients[peer] directly